src/extension/tab/hook.js

```diff
@@ -70,28 +13,63 @@ export function installHook(window, devToolsVersion) {
     },
   });
 
-  // XXX this is a patch to back support previous versions of Apollo Client
-  // at somepoint we should remove this.
-  // the newer version has the client connecting to the hook, not the other
-  // way around that it currently does
-  let interval;
-  let count = 0;
+  function handleActionHookForDevtools() {
+    window.postMessage({
+      message: 'action-hook-fired',
+      to: 'tab:background:devtools',
+    });
+  }
+  
   function findClient() {
-    // only try for 10seconds
-    if (count++ > 10) clearInterval(interval);
-    if (!!window.__APOLLO_CLIENT__) {
-      hook.ApolloClient = window.__APOLLO_CLIENT__;
-      hook.ApolloClient.__actionHookForDevTools(
-        ({
-          state: { queries, mutations },
-          dataWithOptimisticResults: inspector,
-        }) => {
-          hook.actionLog.push({ queries, mutations, inspector });
-        },
-      );
-      clearInterval(interval);
+    let interval;
+    let count = 0;
+  
+    function initializeDevtoolsHook() {
+      if (count++ > 10) clearInterval(interval);
+      if (!!window.__APOLLO_CLIENT__) {
+        hook.ApolloClient = window.__APOLLO_CLIENT__;
+        hook.ApolloClient.__actionHookForDevTools(handleActionHookForDevtools);
+        hook.getQueries = () => hook.ApolloClient.queryManager.queries();
+        hook.getMutations = () => hook.ApolloClient.queryManager.mutationStore.getStore();
+        hook.getCache = () => hook.ApolloClient.cache.extract(true);
+  
+        clearInterval(interval);
+      }
     }
+    
+    interval = setInterval(initializeDevtoolsHook, 1000);
   }
 
-  interval = setInterval(findClient, 1000);
+  // Attempt to find the client on a 1-second interval for 10 seconds max
+  findClient();
+
+  window.addEventListener('message', ({ data }) => {
+    if (data?.message === 'devtools-initialized') {
+      if (hook.ApolloClient) {
+        window.postMessage({
+          message: 'create-devtools-panel',
+          to: 'tab:background:devtools',
+          payload: JSON.stringify({
+            queries: hook.getQueries(),
+            mutations: hook.getMutations(),
+            cache: hook.getCache(),
+          }),
+        });
+      }
+    }
+
+    if (data?.message === 'request-update') {
+      window.postMessage({
+        message: 'update',
+        to: 'tab:background:devtools',
+        payload: JSON.stringify({
```

Comment:
It shouldn't be necessary to `JSON.stringify` the payload, since `postMessage` can serialize basic objects. If there's something that `postMessage` can't handle, I wouldn't trust `JSON.stringify` to handle it any better.

Do you recall why this stringification helped?

Reply:
I still think it would be good to find a way to avoid `JSON.stringify` if possible!

Reply:
I missed removing these earlier. But it look like there's an issue with cloning the following:
<img width="553" alt="Screen Shot 2020-08-11 at 10 02 59 PM" src="https://user-images.githubusercontent.com/5232812/89967048-7e0afb80-dc1e-11ea-98f9-677765417f50.png">

Is that expected? I might be doing something wrong here.